pkg/osinfo/layer_scanner.go

```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package osinfo
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	gzExt        = ".gz"
+	OSDebian     = "debian"
+	OSFedora     = "fedora"
+	OSAlpine     = "alpine"
+	OSDistroless = "distroless"
+)
+
+type LayerScanner struct{}
+
+func (loss *LayerScanner) OSType(layerPath string) (ostype string, err error) {
+	osrelease, err := loss.OSReleaseData(layerPath)
+	if err != nil {
+		if _, ok := err.(ErrFileNotFoundInTar); ok {
+			return "", nil
+		}
+		if strings.Contains(err.Error(), "file not found") {
+			return "", nil
+		}
+		return "", errors.Wrap(err, "reading os release")
+	}
+
+	if osrelease == "" {
+		return "", nil
+	}
+
+	if strings.Contains(osrelease, "NAME=\"Debian GNU") {
+		logrus.Infof("Scan of container layers found %s base image", OSDebian)
+		return OSDebian, nil
+	}
+
+	if strings.Contains(osrelease, "NAME=\"Ubuntu\"") {
+		return OSDebian, nil
+	}
+
+	if strings.Contains(osrelease, "NAME=\"Fedora Linux\"") {
+		return OSFedora, nil
+	}
+
+	if strings.Contains(osrelease, "NAME=\"Alpine Linux\"") {
+		return OSAlpine, nil
+	}
+
+	if strings.Contains(osrelease, "PRETTY_NAME=\"Distroless") {
+		return OSDistroless, nil
+	}
+	return "", nil
+}
+
+// CanHandle looks at an image tarball and checks if it
+// looks like a debian filesystem
+func (loss *LayerScanner) OSReleaseData(layerPath string) (osrelease string, err error) {
+	f, err := os.CreateTemp("", "os-release-")
+	if err != nil {
+		return osrelease, errors.Wrap(err, "creating temp file")
+	}
+	defer f.Close()
+	defer os.Remove(f.Name())
+
+	destPath := f.Name()
+	if err := loss.extractFileFromTar(layerPath, "etc/os-release", destPath); err != nil {
+		return "", errors.Wrap(err, "extracting os-release from tar")
+	}
+	if err != nil {
+		return osrelease, err
+	}
+	data, err := os.ReadFile(destPath)
+	if err != nil {
+		return osrelease, errors.Wrap(err, "reading osrelease")
+	}
+	return string(data), nil
+}
+
+type ErrFileNotFoundInTar struct{}
+
+func (e ErrFileNotFoundInTar) Error() string {
+	return "file not found in tarball"
+}
+
+// extractFileFromTar extracts filePath from tarPath and stores it in destPath
+func (loss *LayerScanner) extractFileFromTar(tarPath, filePath, destPath string) error {
```

Comment:
How about adding this functionality to release-utils or using: https://github.com/kubernetes-sigs/release-utils/blob/3c7b38877ddf515f34dca7922e094f374ea21187/tar/tar.go#L175-L177 ?

Reply:
I looked at the tar package and did not use it because it is missing some features needed here (non-gzipped tarballs, following symlinks, paths starting with dots (`./here`). But I'm planning to add those features to release-utils and remove most of the tar code from this repo.